server/src/main/java/io/spine/server/bus/Bus.java

```diff
@@ -317,14 +319,61 @@ public boolean hasListener(Listener<E> listener) {
      *
      * @param envelopes the envelopes to post
      * @param observer  the observer to be notified of the operation result
-     * @see #dispatch(MessageEnvelope)
+     * @see #dispatch(SignalEnvelope)
      */
+    @SuppressWarnings("ProhibitedExceptionThrown") // Rethrow a caught exception.
     private void doPost(Iterable<E> envelopes, StreamObserver<Ack> observer) {
-        for (E message : envelopes) {
-            queue.add(message, observer);
+        for (E envelope : envelopes) {
+            SignalId signalId = envelope.id();
+            observer.onNext(acknowledge(signalId));
+            onDispatchingStarted(signalId);
+            try {
+                dispatch(envelope);
+            } catch (Throwable t) {
+                _error().withCause(t)
+                        .log("Error when dispatching %s[ID: %s].",
+                             envelope.messageClass(),
+                             signalId);
+                throw t;
+            } finally {
+                onDispatched(signalId);
```

Comment:
I'm not sure if we have to do the same thing (e.g. delivery) if dispatching fails.

Reply:
As discussed vocally, we must tell that the dispatching is completed to set the potentially "waiting" listeners free.